File: cpp/dijkstra.cpp
// dijkstra_benchmark.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>
#include <limits>
#include <chrono>
#include <algorithm>

using namespace std;
using ll = long long;
using Clock = chrono::high_resolution_clock;
const ll INF = numeric_limits<ll>::max() / 4;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cout << "Dijkstra Benchmark (directed weighted graph)\n";
    cout << "Input format:\n";
    cout << "n m\n";
    cout << "u1 v1 w1\n";
    cout << "u2 v2 w2\n";
    cout << "... (m edges, 0-based node indices)\n";
    cout << "s   (source node, 0-based)\n\n";

    int n, m;
    if (!(cin >> n >> m) || n <= 0 || m < 0) {
        cerr << "Invalid n or m.\n";
        return 1;
    }

    vector<vector<pair<int,int>>> adj(n);
    for (int i = 0; i < m; ++i) {
        int u, v;
        long long w;
        if (!(cin >> u >> v >> w)) {
            cerr << "Invalid edge input at line " << (i+1) << ". Expect: u v w\n";
            return 1;
        }
        if (u < 0 || u >= n || v < 0 || v >= n) {
            cerr << "Edge endpoints out of range: " << u << " " << v << "\n";
            return 1;
        }
        // store weight as int if small; cast to int for storage
        adj[u].push_back({v, static_cast<int>(w)});
        // If you want an undirected graph, uncomment the next line:
        // adj[v].push_back({u, static_cast<int>(w)});
    }

    int src;
    if (!(cin >> src) || src < 0 || src >= n) {
        cerr << "Invalid source node.\n";
        return 1;
    }

    // Prepare containers
    vector<ll> dist(n, INF);
    vector<int> parent(n, -1);

    // Start timing Dijkstra (only algorithm execution)
    auto t0 = Clock::now();

    // Min-heap: (distance, node)
    using P = pair<ll,int>;
    priority_queue<P, vector<P>, greater<P>> pq;

    dist[src] = 0;
    pq.push({0, src});

    while (!pq.empty()) {
        auto [d, u] = pq.top();
        pq.pop();
        if (d != dist[u]) continue; // stale entry

        for (const auto &e : adj[u]) {
            int v = e.first;
            int w = e.second;
            ll nd = d + (ll)w;
            if (nd < dist[v]) {
                dist[v] = nd;
                parent[v] = u;
                pq.push({nd, v});
            }
        }
    }

    // End timing
    auto t1 = Clock::now();
    chrono::duration<double> elapsed = t1 - t0;

    // Output distances and paths
    cout << "\nDijkstra results from source node " << src << ":\n";
    for (int v = 0; v < n; ++v) {
        if (dist[v] == INF) {
            cout << "Node " << v << ": unreachable\n";
        } else {
            cout << "Node " << v << ": distance = " << dist[v] << ", path = ";
            // reconstruct path
            vector<int> path;
            int cur = v;
            while (cur != -1) { path.push_back(cur); cur = parent[cur]; }
            reverse(path.begin(), path.end());
            for (size_t i = 0; i < path.size(); ++i) {
                if (i) cout << " -> ";
                cout << path[i];
            }
            cout << "\n";
        }
    }

    cout << "\nDijkstra execution time: " << elapsed.count() << " seconds\n";

    return 0;
}

File: Java/Dijkstra.java
// Main.java
import java.io.*;
import java.util.*;

public class Main {
    static class Edge {
        int to;
        int weight;
        Edge(int to, int weight) { this.to = to; this.weight = weight; }
    }

    static class Node implements Comparable<Node> {
        int v;
        long dist;
        Node(int v, long dist) { this.v = v; this.dist = dist; }
        public int compareTo(Node o) {
            return Long.compare(this.dist, o.dist);
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        // Read n and m
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        if (n <= 0 || m < 0) {
            System.err.println("Invalid n or m.");
            return;
        }

        // Build adjacency list
        ArrayList<Edge>[] adj = new ArrayList[n];
        for (int i = 0; i < n; ++i) adj[i] = new ArrayList<>();

        for (int i = 0; i < m; ++i) {
            String line = br.readLine();
            if (line == null) {
                System.err.println("Insufficient edge lines.");
                return;
            }
            st = new StringTokenizer(line);
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            if (u < 0 || u >= n || v < 0 || v >= n) {
                System.err.println("Edge endpoints out of range: " + u + " " + v);
                return;
            }
            adj[u].add(new Edge(v, w));
            // If you want undirected graph, also add:
            // adj[v].add(new Edge(u, w));
        }

        // Read source
        String sLine = br.readLine();
        if (sLine == null) {
            System.err.println("Missing source node.");
            return;
        }
        int src = Integer.parseInt(sLine.trim());
        if (src < 0 || src >= n) {
            System.err.println("Source out of range.");
            return;
        }

        final long INF = Long.MAX_VALUE / 4;
        long[] dist = new long[n];
        int[] parent = new int[n];
        Arrays.fill(dist, INF);
        Arrays.fill(parent, -1);

        // Run and time only Dijkstra
        long start = System.nanoTime();
        PriorityQueue<Node> pq = new PriorityQueue<>();
        dist[src] = 0;
        pq.add(new Node(src, 0));

        while (!pq.isEmpty()) {
            Node cur = pq.poll();
            int u = cur.v;
            long d = cur.dist;
            if (d != dist[u]) continue; // stale
            for (Edge e : adj[u]) {
                int v = e.to;
                long nd = d + (long)e.weight;
                if (nd < dist[v]) {
                    dist[v] = nd;
                    parent[v] = u;
                    pq.add(new Node(v, nd));
                }
            }
        }
        long end = System.nanoTime();

        double elapsed = (end - start) / 1e9;

        // Output distances and paths
        System.out.println("Dijkstra results from source node " + src + ":\n");
        for (int v = 0; v < n; ++v) {
            if (dist[v] == INF) {
                System.out.println("Node " + v + ": unreachable");
            } else {
                System.out.print("Node " + v + ": distance = " + dist[v] + ", path = ");
                // reconstruct path
                ArrayList<Integer> path = new ArrayList<>();
                int cur = v;
                while (cur != -1) {
                    path.add(cur);
                    cur = parent[cur];
                }
                Collections.reverse(path);
                for (int i = 0; i < path.size(); ++i) {
                    if (i > 0) System.out.print(" -> ");
                    System.out.print(path.get(i));
                }
                System.out.println();
            }
        }

        System.out.printf("%nDijkstra execution time: %.9f seconds%n", elapsed);
    }
}
